src/provider/repository/stackablerepository.rs

```diff
@@ -0,0 +1,197 @@
+            self.get_repo_metadata().await?;
+        }
+        if let Some(content) = &self.content {
+            let parcels = &content.parcels;
+            if let Some(product) = parcels.get(&package.product) {
+                // product exists in repo
+                if let Some(version) = product.get(&package.version) {
+                    // found our package
+                    return Ok(version.clone());
+                }
+            };
+        }
+        Err(PackageNotFound {package})
+    }
+
+    pub async fn download_package(&mut self, package: &Package, target_path: PathBuf) -> Result<(), StackableError> {
+        if self.content.is_none() {
+            let _content = self.get_repo_metadata();
+        }
+
+        let stackable_package = self.get_package(package.clone()).await?;
+        let download_link = Url::parse(&stackable_package.link)?;
+        let response = reqwest::get(download_link).await?;
+
+        let mut content =  Cursor::new(response.bytes().await?);
+
+        let mut out = File::create(target_path.join(package.get_file_name()))?;
+        copy(&mut content, &mut out)?;
+        Ok(())
+    }
+
+    // TODO: implement caching based on version of metadata
+    async fn get_repo_metadata(&mut self) -> Result<RepositoryContent, StackableError> {
+        trace!("entering get_repo_metadata");
+        let mut metadata_url = self.base_url.clone();
+
+        // TODO: add error propagation
+        // path_segments_mut returns () in an error case, not sure how to handle this
+        metadata_url
+            .path_segments_mut()
+            .expect("")
+            .push("metadata.json");
+
+        debug!("Retrieving repository metadata from {}", metadata_url);
+
+        let repo_data = reqwest::get(metadata_url).await?.json::<RepoData>().await?;
+
+        debug!("Got repository metadata: {:?}", repo_data);
+
+        let mut parcels: HashMap<String, HashMap<String, StackablePackage>> = HashMap::new();
+        for (product, versions) in repo_data.parcels {
+            let mut versionlist = HashMap::new();
+            for version in versions {
+                versionlist.insert(
+                    version.version.clone(),
+                    StackablePackage {
+                        product: product.clone(),
+                        version: version.version,
+                        link: self.resolve_url(version.path.clone())?,
+                        hashes: version.hashes.clone()
+                    },
+                );
+            }
+            parcels.insert(product, versionlist);
+        }
+        let repo_content: RepositoryContent = RepositoryContent {
+            version: repo_data.version,
+            parcels,
+        };
+        self.content = Some(repo_content.clone());
+        Ok(repo_content)
+    }
+
+    fn resolve_url(&self, path: String) -> Result<String, StackableError> {
+        if let Result::Ok(absolute_link) = Url::parse(&path) {
```

Comment:
`absolute_link` is not being used here. If that's correct you can just use `_`

Reply:
I've changed that, added a small comment and a test case.